code/API_definitions/qod-api.yaml

```diff
@@ -204,6 +204,153 @@ paths:
           $ref: "#/components/responses/Generic503"
       security:
         - apiKey: []
+  /qos-profiles:
+    get:
+      tags:
+        - QoS Profiles
+      summary: "Get All QoS Profiles"
+      operationId: getQosProfiles
+      parameters:
+        - name: id
+          in: query
+          schema:
+            type: string
+          required: false
+        - name: status
+          in: query
+          schema:
+            $ref: '#/components/schemas/QosProfileStatusEnum'
+          required: false
+        - name: targetMinimumUpstreamRate
```

Comment:
The intended serialisation for these complex objects is not clear to me. For example, if I am searching for a profile that supports a minimum upstream data rate of exactly 10 Mb/s, would the serialisation be:

`GET /qos-profiles?targetMinimumUpstreamRate={"value": "10", "unit": "Mbps"}`

I don't believe OAS defined serialisation in that way, so this should be documented.

Also, syntax for filtering is defined in the [CAMARA design guidelines](https://github.com/camaraproject/WorkingGroups/blob/main/Commonalities/documentation/API-design-guidelines.md#83-filtering), so we should try to follow those. So I should be able to search for all profiles that have a parameter higher or lower than a specified value, which is when filtering becomes much more useful. e.g.

`GET /qos-profiles?priority.gte=5`

which would mean "return all profiles with a priority greater or equal to 5". But how to do this for complex objects?

Maybe: `GET /qos-profiles?targetMinimumUpstreamRate.gte={"value": "10", "unit": "Mbps"}`

I agree with @jlurien that this could all get complex to implement for complex objects, but first lets be clear about how filtering is meant to work and, if different from the mechanism documented in the design guidelines, why is a different approach being proposed and do we need to update the guidelines?

Personally, I think filtering would be useful here, but it is complicated by the use of complex objects in the API definition.

Reply:
Unless we have clear requirements for the convenience or necessity of filters, I would go for a simple GET without filters for the first iteration, so we can focus here on how to define a profile.

Reply:
@RandyLevensalor
As mentioned above, the CAMARA API Design Guidelines have a well-defined syntax for [filtering](https://github.com/camaraproject/WorkingGroups/blob/main/Commonalities/documentation/API-design-guidelines.md#83-filtering).

So:
`GET /qos-profiles?priority=5`
means return all profiles where `priority` is exactly equal to 5, whereas:
`GET /qos-profiles?priority.gte=5`
means return all profiles where `priority` is greater than or equal to 5

If you do not intend this behaviour, you really need to justify why, for otherwise API implementors will have to implement different filtering implementations for different APIs, which they really don't want to do. That's why we have guidelines.

And if you do intend this behaviour, you need to decide how you want it to work for complex objects according to the [OAS parameter serialisation](https://swagger.io/docs/specification/serialization/) rules. I think you may need to add `style: deepObject` if you want code generators to support this, but even then that only works for equalities. As the design guidelines don't consider filtering on complex object values, these would also require an update.

@jlurien 
I think filtering on `name` (the "contains" variant) and `status` is required (why would I want to see `Inactive` or `Deprecated` profiles?). But these are simple strings and the filtering syntax is clear. The value of filtering on other parameters is less clear and, whilst the filtering syntax for these complex objects is also unclear, I'd agree to remove filtering for those parameters for now.

Reply:
@eric-murray

The example you provided around highlights why I think applying numeric operators to many of the qos profile fields is confusing and ambiguous.

If we were to interpert this as a puerly numeric operator, then `GET /qos-profiles?priority.gte=5` would return QoS Profiles with a priority no higher than 5.  For instance a priority of 6 would be returned and that would have a lower priority than 5.

When I read:

> return all profiles where priority is greater than or equal to 5

I would expect that to return higher priority profiles, but a lower numeric priority value.  So a priority 2.5 would be a higher priority than 5.

What about updating the description of the query to say something like
"The get query will return all profiles to can meet the stated filter criteria."

Then for each attribute, we can provide example of what types of results would be expected.  For instance, priority would be the stated priority or a higher priority.  So it would actually be equivalent to the lte numeric operator.

This doesn't address the issues with complex data types.  I'd like to get to consensus on this first.

Reply:
@RandyLevensalor 
But that is how developers will expect filtering to work. If they ask for all profiles with a priority value of 5 or higher, they will expect to see QoS profiles with priority = 6, but not with priority = 4.

Part of the problem is that you have defined `priority` as `The priority of this QoS Profile`, with no indication as to whether a higher number represents a higher or lower scheduling priority. 3GPP's use of lower numbers for higher scheduling priority is just a convention (albeit a common one) and one with which the API consumer may not be familiar. If it is explained that lower priority numbers mean a higher scheduling priority, then it will be obvious that filtering for priority values less than or equal to the lowest scheduling priority acceptable will return a list suitable QoS profiles.

But expecting `GET /qos-profiles?priority.gte=5` to return QoS profiles with a priority value less than or equal to 5 is so counter-intuitive that it should not be implemented.

Anyway, my proposal to @jlurien was only to filter on the string fields `name` and `status`. That remains my proposal for now.

Reply:
@eric-murray  I'm fine with limiting filters. `status` makes sense to discard profiles others than active. About `name`, if that is the friendly unique identifier of the profile, probably makes sense to use it as path parameter instead of the `id`,  as `GET /qos-profiles/{name}`. We may not need both `id` and `name` as identifiers.

Reply:
@jlurien My understanding was that `name` does not need to be unique, but `id` does. So `name` is a (fairly short) free text field that the API provider can use to indicate to the API caller for which use cases a specific profile might be useful. But there could be several similar profiles that are given the same name but differ in only minor ways.

For example, an API provider might include the term `conversational_video` in multiple profile names, and the API caller can find them using:
`GET /qos-profiles/?name~=conversational_video`

An API provider who thinks it important that the API caller know what QCI they are getting for their money can include that in the `name`:
`GET /qos-profiles/?name~=qci4`

So searching on `name` makes sense, but not requiring that it be unique also makes sense.

Reply:
@eric-murray @sfnuser @jlurien 

What if we change the Duration to be an int64 that is in microseconds.  This can encompass several years and should be small enough to handle future requirements around jitter.  Conversions would need to be done to make these human readable, but it would allow for a consistent and searchable value across all time fields.

For Rate, that can be simplified to bits per second (bps) as a int64.

Reply:
One more note.  Using a name with some constraints is consistent with the way amazon implemented their EC2 instance types.  EC2 instance sizing is a rough analogy for the QoSProfiles.  https://aws.amazon.com/ec2/instance-types.  So this pattern should be familiar to many application developers.

Reply:
> Anyway, my proposal to @jlurien was only to filter on the string fields `name` and `status`. That remains my proposal for now.

I second this proposal. The content of this PR should be to get the QoS Profile structure right. 
Instead of complex searches I would envision a new method which allows a developer to define his connectivity needs and they will get a list of matching profiles (or even only the one which is above and nearest to their needs).